models/dsn_helpers.go

```diff
@@ -0,0 +1,154 @@
+	// TODO TLSConfig: "true", https://jira.percona.com/browse/PMM-1727
+	// TODO Other parameters?
+
+	cfg := mysql.NewConfig()
+	cfg.User = pointer.GetString(exporter.Username)
+	cfg.Passwd = pointer.GetString(exporter.Password)
+	cfg.Net = "tcp"
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	cfg.Addr = net.JoinHostPort(host, strconv.Itoa(int(port)))
+	cfg.Timeout = 1 * time.Second
+	cfg.DBName = db
+
+	// QAN code in pmm-agent uses reform which requires those fields
+	cfg.ClientFoundRows = true
+	cfg.ParseTime = true
+
+	return cfg.FormatDSN()
+}
+
+// DSNforPostgreSQL generates Postgres DSN string from service and exporter.
+func DSNforPostgreSQL(service *Service, exporter *Agent) string {
+	q := make(url.Values)
+	q.Set("sslmode", "disable") // TODO: make it configurable
+	q.Set("connect_timeout", "1")
+
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	username := pointer.GetString(exporter.Username)
+	password := pointer.GetString(exporter.Password)
+
+	u := &url.URL{
+		Scheme:   "postgres",
+		Host:     net.JoinHostPort(host, strconv.Itoa(int(port))),
+		Path:     "postgres",
+		RawQuery: q.Encode(),
+	}
+	switch {
+	case password != "":
+		u.User = url.UserPassword(username, password)
+	case username != "":
+		u.User = url.User(username)
+	}
+
+	return u.String()
+}
+
+// DSNforMongoDB generates MongoDB DSN string from service and exporter.
+func DSNforMongoDB(service *Service, exporter *Agent) string {
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	username := pointer.GetString(exporter.Username)
+	password := pointer.GetString(exporter.Password)
+
+	u := &url.URL{
+		Scheme: "mongodb",
+		Host:   net.JoinHostPort(host, strconv.Itoa(int(port))),
+	}
+	switch {
+	case password != "":
+		u.User = url.UserPassword(username, password)
+	case username != "":
+		u.User = url.User(username)
+	}
+
+	return u.String()
+}
+
+// FindDSNByServiceID resolves DSN by service id.
+func FindDSNByServiceID(q *reform.Querier, serviceID, db string) (string, error) {
+	var result string
+
+	svc, err := FindServiceByID(q, serviceID)
+	if err != nil {
+		return "", err
+	}
+
+	pmmAgents, err := FindPMMAgentsForService(q, serviceID)
+	if err != nil {
+		return "", err
+	}
+
+	if len(pmmAgents) != 1 {
+		return "", status.Errorf(codes.FailedPrecondition, "Couldn't resolve dsn, as there should be only one pmm-agent")
+	}
+
+	pmmAgentID := pmmAgents[0].AgentID
+	var agentType AgentType
+	switch svc.ServiceType {
+	case MySQLServiceType:
+		agentType = MySQLdExporterType
+	case MongoDBServiceType:
+		agentType = MongoDBExporterType
+	case PostgreSQLServiceType:
+		agentType = PostgresExporterType
+	default:
+		return "", status.Errorf(codes.FailedPrecondition, "Couldn't resolve dsn, as service is unsupported")
+	}
+
+	exporters, err := FindAgentsByPmmAgentIDAndAgentType(q, pmmAgentID, agentType)
```

Comment:
Why do we  look for agent by pmm-agent and agent type instead of finding by service id and agent type?

Reply:
There might be another mysql agent for another service.

Reply:
Because we need both `service` and `exporter` model in the code below, to make `DSN`.

Reply:
I mean we can find agent by service, not by pmm-agent. But the best way to find agent is by pmm-agent and service.

Reply:
So can you suggest function notation or something like this? Some piece of code would be great as I don't understand where the problem is.

Reply:
Ok, look.

We have 3 mysqld exporters, 2 MySQL services and 2 PMM-agents.
mysqld-exporter-1 and mysqld-exporter-2 runs on pmm-agent-1
mysqld-exporter-3 runs on pmm-agent-2
mysqld-exporter-1 works for mysql-service-1
mysqld-exporter-2 and mysqld-exporter-3  work for mysql-service-2

So we want to run explain action for mysql-service-2. How should we do it?
Now you logic might get dns for mysqld-exporter-1 and mysql-service-2 and it won't work.

Reply:
@BupycHuk Check, please.